enclave_build: Image manager trait

raulmldv opened the pull request:
Image Manager

The image manager trait exposes the operations required for building the EIF. The decision wether to fetch the image details from the local OCI storage or if they should be pulled from a registry is implemented as a failover logic that will allow the build to continue even if the image is missing locally, the storage is corrupt or missing.

The same manager trait will be used for building EIFs from Docker images but will be adapted in a future PR with the implied workflow changes in `lib.rs`.


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,144 @@
+
+        if let Some(storage) = local_storage {
+            // Try to fetch the image from the storage
+            let image_details = storage.fetch_image_details(&image_name).map_err(|err| {
+                // Log the fetching error
+                eprintln!("{:?}", err);
+                err
+            });
+
+            // If the fetching failed, pull it from remote and store it
+            return image_details.or(self.fetch_and_try_store_image(&image_name).await);
```
```rust
            return storage
                .fetch_image_details(&image_name)
                .map_err(|err| {
                    // Log the fetching error
                    eprintln!("{:?}", err);
                    err
                })
                .or(self.fetch_and_try_store_image(&image_name).await);
```

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,144 @@
+        };
+
+        // Try to create/read the storage
+        let storage = match OciStorage::new(&root_path) {
+            Ok(manager) => Some(manager),
+            Err(err) => {
+                // If the storage could not be created, log the error
+                eprintln!("{:?}", err);
+                None
+            }
+        };
```
I think we can use `map_err(...).ok()` here.

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,144 @@
+    /// it pulls the image, stores it (if the 'storage' field is not None) and returns its metadata.
+    ///
+    /// If the pull succeeded but the store operation failed, it returns the pulled image metadata.
+    async fn get_image_details(&mut self, image_name: &str) -> Result<ImageDetails> {
```
Why do we need a mutable reference? From the function name, I would imagine self is not mutated. I think the problem comes from the fact that we are doing more things than we should in a single function, maybe we can refactor a bit (do the storing separately). Is that doable?

raulmldv replied:
The idea of this image manager is to hide the storage implementation while allowing the EIF to be built even if we don't have the image stored. We have to mutate the storage at some point and I think it complicates things if we move it higher in the workflow.

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,144 @@
+            return image_details.or(self.fetch_and_try_store_image(&image_name).await);
+        }
+
+        self.fetch_and_try_store_image(&image_name).await
```
Can we refactor this a bit so that we call `fetch_and_try_store_image` only once?

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,144 @@
+        match (cmd, env, entrypoint) {
+            (Some(cmd), Some(env), _) => Ok((cmd.to_vec(), env.to_vec())),
+            (_, Some(env), Some(entrypoint)) => Ok((entrypoint.to_vec(), env.to_vec())),
+            (_, _, Some(entrypoint)) => Ok((entrypoint.to_vec(), Vec::<String>::new())),
```
nit: no need for the explicit `String`.

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,144 @@
+    /// Extracts from the image and returns the CMD and ENV expressions (in this order).
+    ///
+    /// If there are no CMD expressions found, it tries to locate the ENTRYPOINT command.
+    fn extract_image(&mut self) -> Result<(Vec<String>, Vec<String>)> {
```
Same comment here: the method should take an immutable reference judging by its name.

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -6,8 +6,25 @@ use crate::image::ImageDetails;
 use crate::storage::OciStorage;
 use crate::{EnclaveBuildError, Result};
 
+use tempfile::NamedTempFile;
```
Can we use the one from vmm-sys-util?

raulmldv replied:
For now we match the `load` operation from the Docker workflow. I suggest we change it everywhere in another PR and remove the `tempfile` dependency altogether. Right now I would have to modify too many modules that are not related to these changes.

lauralt replied:
Got it, thanks

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -6,8 +6,25 @@ use crate::image::ImageDetails;
 
+/// Trait which provides an interface for handling images.
+pub trait ImageManager {
+    fn image_name(&self) -> &String;
```
`&String` is quite unusual, can we replace with `&str`?

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -223,3 +223,43 @@ impl ImageManager for OciImageManager {
     }
 }
+
+#[cfg(test)]
```
The unit test shouldn't be in a separate commit than the code it is testing.

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,261 @@
+        // The image is not stored, so try to pull and then store it
+        let image_data = crate::pull::pull_image_data(image_name).await?;
+
+        // If the store operation failed, still return the image details
```
This comment should be after attempting the store, it is confusing otherwise.

raulmldv commented:
I addressed the comments and redesigned the interface for the `ImageManager` trait. I applied suggested renaming and deleted some functions which imply modifying the Docker workflow. I decided to leave those changes to the following PR and focus on OCI for this one.

The trait is now simplified (removed build and pull operations) and no longer takes mutable references for the manager. To achieve that I decided to keep the image details at the image manager level (instead of keeping the storage). The details will be fetched on manager creation either from the local storage or pulled from the registry.

belveryin commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,251 @@
+use crate::storage::OciStorage;
+use crate::{EnclaveBuildError, Result};
+
+/// Trait which provides an interface for handling images.
```
Nit: maybe this could be rephrased to "handling a particular image". The first time I read it I thought it would manage a collection of images.

belveryin commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,251 @@
+            Ok(details) => Ok(details),
+            Err(err) => {
+                // Log error
+                eprintln!("Fetching from storage failed: {}", err);
```
Nit: if we fail in line 72, the error is logged twice: in line 74 and also here, in 88.

lauralt replied:
I think we can keep the error logging only here.

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,178 @@
+    /// If the pull succeeded but the store operation failed, it returns the pulled image metadata.
+    async fn fetch_image_details(
+        image_name: &str,
+        mut storage: Option<OciStorage>,
```
Also, why not have the parameter as `Option<&mut OciStorage>` so that you avoid doing the first instruction in the function?

raulmldv replied:
As fetching from the storage and storing would both mutate the storage, later we need a binding instead of the first line.

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,178 @@
+                }
+
+                // Get the image metadata from the pulled struct
+                let image_details = ImageDetails::build_details(image_name, &image_data)?;
+
+                Ok(image_details)
```
nit: -> `ImageDetails::build_details(image_name, &image_data)`

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,178 @@
+/// Adds the default ":latest" tag to an image if it is untagged
+fn normalize_tag(image_name: &str) -> String {
+    let name = image_name.to_string();
+    match name.contains(':') {
```
Is any name that contains at least one ":" a valid tagged image? Is it valid to have "::" for example?

raulmldv replied:
Found a tag parser in `oci_distribution`

lauralt commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,180 @@
+                err
+            })
+        } else {
+            Err(EnclaveBuildError::OciStorageNotFound(
```
Do we ignore this error later? I'll try to refactor a bit this method.

raulmldv replied:
We were logging it before in line 78 but decided to only log once. I think I misunderstood and removed that.

raulmldv commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,180 @@
+            // Try to fetch the image from the storage
+            storage.fetch_image_details(image_name).map_err(|err| {
+                // Log the fetching error
+                eprintln!("{:?}", err);
```
I will change the logging.

petreeftime commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,180 @@
+        // If the fetching failed, pull it from remote and store it
+        match image_details {
+            Ok(details) => Ok(details),
+            Err(_) => {
```
I think this err might be catching a few too many things. Yes, it might be that the image is not already in storage, but it could also be that we don't have access to the storage (ie. permission issues) or that the folder doesn't exist and it might have to be created either by this library or, I think, preferably by the user.

petreeftime commented on enclave_build/src/image.rs:
```diff
@@ -53,6 +54,40 @@ impl ImageDetails {
+            .as_ref()
+            .ok_or(EnclaveBuildError::ConfigError)?;
+
+        let cmd = config_section.cmd();
+        let env = config_section.env();
+        let entrypoint = config_section.entrypoint();
+
+        // If no CMD instructions are found, try to locate an ENTRYPOINT command
+        let (cmd, env) = match (cmd, env, entrypoint) {
+            (Some(cmd), Some(env), _) => Ok((cmd.to_vec(), env.to_vec())),
+            (_, Some(env), Some(entrypoint)) => Ok((entrypoint.to_vec(), env.to_vec())),
+            (_, _, Some(entrypoint)) => Ok((entrypoint.to_vec(), Vec::new())),
+            (_, _, _) => Err(EnclaveBuildError::ExtractError(
+                "Failed to locate ENTRYPOINT".to_string(),
+            )),
+        }?;
```
More readable as
```
let env = config_section.env().ok_or_default();
let cmd = config_section.cmd()
    .ok_or(|| config_section.entrypoint())
    .map_err(...)?;
```

petreeftime commented on enclave_build/src/image_manager.rs:
```diff
@@ -0,0 +1,180 @@
+
+    #[cfg(target_arch = "x86_64")]
+    const SAMPLE_IMAGE: &str =
+        "667861386598.dkr.ecr.us-east-1.amazonaws.com/enclaves-samples:vsock-sample-server-x86_64";
```
Let's publish some empty test images into our Public ECR so that anyone can pull these instead.
